packages/shared/src/components/squads/SquadJoinButton.tsx

```diff
@@ -0,0 +1,86 @@
+import React, { ReactElement, useMemo } from 'react';
+import classNames from 'classnames';
+import { useMutation, useQueryClient } from 'react-query';
+import { Squad } from '../../graphql/sources';
+import { Button } from '../buttons/Button';
+import { useAuthContext } from '../../contexts/AuthContext';
+import { useToastNotification } from '../../hooks/useToastNotification';
+import { useLeaveSquad, useJoinSquad } from '../../hooks';
+import { labels } from '../../lib';
+
+type SquadJoinProps = {
+  className?: string;
+  squad: Squad;
+  joinText?: string;
+  leaveText?: string;
+};
+
+export const SquadJoinButton = ({
+  className,
+  squad,
+  joinText = 'Join squad',
+  leaveText = 'Leave squad',
+}: SquadJoinProps): ReactElement => {
+  const queryClient = useQueryClient();
+  const { displayToast } = useToastNotification();
+  const { squads, user, showLogin } = useAuthContext();
+  const isCurrentMember = useMemo(
+    () => squads?.some(({ id }) => id === squad.id) || false,
+    [squad, squads],
+  );
+
+  const { mutateAsync: joinSquad, isLoading: isJoiningSquad } = useMutation(
```

Comment:
Is the useJoinSquad not already a mutation?
I think ideally that logic all sits in the hook normally

Reply:
Not really it is just a function that is exported that does "bunch of things that are related". Kinda like your comment https://github.com/dailydotdev/apps/pull/1948#discussion_r1266810470 with which I agree.

If we start just re-exporting useMutation wrapped in custom hook we then often return from it something like:
`{ onJoinSquad, isLoading }`

Then we need for example another property `mutation.isDone`, we add that.

Then we need `onSuccess`, we add that to hook props.. and so on..

Pretty fast hook has X props and Y return values.

I think it is much cleaner and we should think of `useMutation` as a wrapper for loading state, success/error callbacks in specific implementation vs hardcoding them inside custom hooks which then become really specific and hard to expand.

Reply:
Alright valid point, but at which you would almost argue why the custom hook then.
For me it's a wrapping place where all related logic happens so we don't have to do it in multiple places.

Tracking, rollbacks, toast etc

Reply:
> Alright valid point, but at which you would almost argue why the custom hook then.

Custom hook here is because we need context and React context to actually get to our tracking function and updating boot. If not we would not need it and it could just be exported.

> Tracking, rollbacks, toast etc

I would argue side effects should not be there because more times those differ from component to component. 

Or we would have to be very specific for what hook is used. Basically each new argument we add to the hook we make it more complex and with more variations on how it works. 

We could thought copy what we are doing `useVotePost` where `mutationHandlers` are exported and reused across different usages of the hooks. 

That way we could still have the function combine all the related logic but side effects can be injected again where needed from predefined functions (or hooks).